enclave/src/polkadex.rs

```diff
@@ -0,0 +1,159 @@
+
+use chain_relay::{storage_proof::StorageProofChecker, Header};
+use codec::Encode;
+use core::hash::Hasher;
+use core::ops::Deref;
+use multibase::Base;
+use polkadex_primitives::PolkadexAccount;
+use sgx_tstd::collections::HashMap;
+use sgx_tstd::hash::Hash;
+use sgx_types::{sgx_epid_group_id_t, sgx_status_t, sgx_target_info_t, SgxResult};
+use sp_runtime::traits::{AccountIdConversion, Header as HeaderT};
+use sp_runtime::ModuleId;
+use sp_std::prelude::*;
+use std::sync::{
+    atomic::{AtomicPtr, Ordering},
+    Arc, SgxMutex, SgxMutexGuard,
+};
+use substratee_stf::{
+    AccountId, Getter, ShardIdentifier, Stf, TrustedCall, TrustedCallSigned, TrustedGetterSigned,
+};
+
+//use std::collections::HashMap;
+// TODO: Fix this import
+use crate::utils::UnwrapOrSgxErrorUnexpected;
+
+static GLOBAL_ACCOUNTS_STORAGE: AtomicPtr<()> = AtomicPtr::new(0 as *mut ());
+
+pub fn verify_pdex_account_read_proofs(
+    header: Header,
+    accounts: Vec<PolkadexAccount>,
+) -> SgxResult<()> {
+    let mut last_account: AccountId = ModuleId(*b"polka/ga").into_account();
+    for account in accounts.iter() {
+        if account.account.prev == last_account.as_ref() {
+            StorageProofChecker::<<Header as HeaderT>::Hashing>::check_proof(
+                header.state_root,
+                account.account.current.as_ref(), // QUESTION: How is this key defined? What about storage prefix?
+                account.proof.to_vec(),
+            )
+            .sgx_error_with_log("Erroneous StorageProof")?;
+
+            last_account = AccountId::from(account.account.next.clone().unwrap());
+        }
+    }
+
+    Ok(())
+}
+
+pub fn create_in_memory_account_storage(accounts: Vec<PolkadexAccount>) -> SgxResult<()> {
+    let accounts_storage = PolkadexAccountsStorage::create(accounts);
+    let storage_ptr = Arc::new(SgxMutex::<PolkadexAccountsStorage>::new(accounts_storage));
+    let ptr = Arc::into_raw(storage_ptr);
+    GLOBAL_ACCOUNTS_STORAGE.store(ptr as *mut (), Ordering::SeqCst);
+    Ok(())
+}
+
+/// Access that pointer
+pub struct PolkadexAccountsStorage {
+    pub(crate) accounts: HashMap<[u8;32], Vec<[u8;32]>>,
+}
+
+impl PolkadexAccountsStorage {
+
+    pub fn from_hashmap(hashmap: HashMap<[u8;32], Vec<[u8;32]>>) -> Self{
+        Self{
+            accounts: hashmap
+        }
+    }
+
+    pub fn create(accounts: Vec<PolkadexAccount>) -> PolkadexAccountsStorage {
+        let mut in_memory_map: PolkadexAccountsStorage = PolkadexAccountsStorage {
+            accounts: HashMap::new(),
+        };
+        for account in accounts {
+            in_memory_map
+                .accounts
+                .insert(account.account.current, account.account.proxies);
+        }
+        in_memory_map
+    }
+
+    pub fn add_main_account(&mut self, acc: [u8;32]) {
+        let vec: Vec<[u8;32]> = Vec::new();
+        self.accounts.insert(acc, vec);
```

Comment:
Why not use the return value of the hashmap? I think an error!(..) log msg in case of trying to add an already existing main account could help to debug